### PR TITLE
AWS Advanced Template support for EL7

### PR DESCRIPTION
**Pull request by lingmann**
Support for EL7 and custom AMI's with AWS Advanced Templates

### CODE REVIEW

**Comment by lingmann** [comment]
@spahl or @cmaloney could I please get a review?

File: gen/__init__.py [hunk 1]
```diff
@@ -48,8 +48,48 @@ def add_roles(cloudconfig, roles):
...
+    * cloud_init_implementation is a string: 'coreos' or 'canonical'
+    '''
+    if cloud_init_implementation == 'canonical':
+        for unit in services:
+            unit_name = unit['name']
+            if 'content' in unit:
+                write_files_entry = {'path': '/etc/systemd/system/{}'.format(unit_name),
+                                     'content': unit['content'],
+                                     'permissions': '0644'}
+                cloudconfig.setdefault('write_files', []).append(write_files_entry)
```
Comment (cmaloney) [inline gen/__init__.py]: Why not just always default this right under `cloud_init_implementation == 'canonical'`? Seems like ti would be easier to read (And I would hope an empty write_files would be fine (Not to mention the odds of us ever having no units with content are very slim))

File: gen/aws/templates/advanced/advanced-master.json [hunk 2]
```diff
@@ -46,39 +46,69 @@
...
+          ]
+        }
+      ]
     }
   },
   "Mappings": {
     "RegionToAmi": {
       "ap-northeast-1": {
-        "stable": "ami-84e0c7ea"
+        "coreos": "ami-84e0c7ea",
```
Comment (cmaloney) [inline gen/aws/templates/advanced/advanced-master.json]: A little odd that we inject all the AMIs into both the templates (Even though just the el7 or just the stable will work). That said, probably good enough.

File: gen/azure/cloud-config.yaml [hunk 3]
```diff
@@ -110,9 +69,6 @@ runcmd:
     - [ rm, -f, /etc/resolv.conf ]
     - [ cp, -p, /tmp/resolv.conf, /etc/resolv.conf ]
     - [ systemctl, start, dcos-docker-install.service ]
-    - [ systemctl, start, dcos-link-env.service ]
```
Comment (cmaloney) [inline gen/azure/cloud-config.yaml]: Just to confirm: moving all these after dcos-config-writer is fine?

File: gen/installer/aws.py [hunk 4]
```diff
@@ -56,11 +56,10 @@
 
 late_services = """- name: dcos-cfn-signal.service
   command: start
+  no_block: true
   content: |
     [Unit]
     Description=AWS Setup: Signal CloudFormation Success
-    After=dcos.target
```
Comment (cmaloney) [inline gen/installer/aws.py]: :+1:

File: gen/installer/aws.py [hunk 5]
```diff
@@ -219,13 +218,17 @@ def gen_supporting_template():
...
-extra_templates = ['aws/dcos-config.yaml', 'coreos-aws/cloud-config.yaml', 'coreos/cloud-config.yaml']
-
-
 def make_advanced_bunch(variant_args, template_name, cc_params):
+    extra_templates = ['aws/dcos-config.yaml', 'aws/templates/advanced/{}'.format(template_name)]
+    if cc_params['os_type'] == 'coreos':
+        extra_templates += ['coreos-aws/cloud-config.yaml', 'coreos/cloud-config.yaml']
+        cloud_init_implementation = 'coreos'
+    else:
+        cloud_init_implementation = 'canonical'
```
Comment (cmaloney) [inline gen/installer/aws.py]: can we have an assert here that `cc_params['os_type']` is equal to what we expect here? (I worry someone goes to add another os_type, doesn't realize they should touch this code, things sort of work for mysterious reasons until one day it all explodes with an unrelated change)

**Comment by cmaloney** [comment]
Generally looks pretty good to me! Definitely like the "Use custom AMI to extend arbitrarily"

File: gen/aws/templates/advanced/advanced-master.json (hunk 2 above)
Comment (lingmann) [inline gen/aws/templates/advanced/advanced-master.json]: Agreed, for future improvements would like to templatize the whole AMI map, but wanted to keep this simple for now and not introduce potential template rendering issues.

File: gen/azure/cloud-config.yaml (hunk 3 above)
Comment (lingmann) [inline gen/azure/cloud-config.yaml]: Yup

File: gen/calc.py [hunk 6]
```diff
@@ -234,6 +234,11 @@ def calculate_adminrouter_auth_enabled(oauth_enabled):
     return oauth_enabled
 
 
+def validate_os_type(os_type):
+    can_be = ['coreos', 'el7']
+    assert os_type in can_be, 'Must be one of {}. Got {}'.format(can_be, os_type)
```
Comment (lingmann) [inline gen/calc.py]: @cmaloney should I include a reference to this under `entry['validate']` in `gen/calc.py` or `gen/aws/calc.py`?

File: gen/__init__.py (hunk 1 above)
Comment (lingmann) [inline gen/__init__.py]: Thanks that seems like a good improvement... done.

File: gen/installer/aws.py (hunk 5 above)
Comment (lingmann) [inline gen/installer/aws.py]: Yeah, I had a thought about that as well... this tells me we should do it. Adjustments made.

File: gen/calc.py (hunk 6 above)
Comment (cmaloney) [inline gen/calc.py]: For now we can drop in tin `gen/aws/calc.py` since azure doesn't follow it. We should eventually move it up another level.
